span-processing-utils/src/main/java/org/hypertrace/config/span/processing/utils/SpanFilterMatcher.java

```diff
@@ -38,18 +45,24 @@ public boolean matchesEnvironment(SpanFilter spanFilter, Optional<String> enviro
   }
 
   public boolean matchesServiceName(SpanFilter spanFilter, String serviceName) {
+    if (spanFilter.hasUnarySpanFilter()) {
+      return true;
+    }
     if (spanFilter.hasRelationalSpanFilter()) {
       return matchesServiceName(spanFilter.getRelationalSpanFilter(), serviceName);
     } else {
       if (spanFilter
           .getLogicalSpanFilter()
           .getOperator()
           .equals(LogicalOperator.LOGICAL_OPERATOR_AND)) {
+        if (hasNoRelationalFilters(spanFilter.getLogicalSpanFilter())) {
+          return true;
```

Comment:
This doesn't necessarily hold. It seems like you're making assumptions about structure here.  e.g. `AND(AND(serviceName = x, otherCondition))` would stop searching as soon as it got inside the first AND.